Watchtower revocation signatures exchange

darosior opened the pull request:
This pull request, based on #306, implements the distribution of revocation signatures to the watchtower (https://github.com/revault/miradord).

The protocol defines a `sig` message per transaction and we leverage this on the watchtower to not allocate fee-bumping reserves before a vault is delegated. In order to do this we share the second-stage (ie the txs spending the Unvault) transactions signatures before sharing the Unvault to implicitly signal that we are about to share the Unvault transaction signature and the watchtower better be telling us not to do it now if it doesn't have sufficient reserve [0].
Of course, this is all simulation as we don't have fee-bumping reserve management implemented on the watchtower yet.
This translates here by having the background signature fetcher thread share the Emergency transaction signatures when it gathered them all (or to do it immediately in `revocationtxs` if we were passed them all at once).

This also adds a few plugins to test some basic spending policies (revault everything, revault nothing, revault according to a maximum per day). Note that by default we set the "revault nothing" policies as by default the watchtower would currently revault everything and this would break most of our tests :).

[0] I think we should not rely on this trick but instead modify the protocol. See https://github.com/revault/practical-revault/issues/109

darosior commented:
Arg, small typo in the CI config to build the watchtower. Should be fixed now.

darosior commented:
Rebased on master

darosior commented:
Ok so the policy test is a bit flaky, can't manage to reproduce locally... It's ready for review nonetheless!

danielabrozzoni commented on tests/test_rpc.py:
```diff
@@ -1108,25 +1108,21 @@ def test_getserverstatus(revault_network, bitcoind):
-        # ...And well, they're dead
         for watchtower in res["watchtowers"]:
-            assert not watchtower["reachable"]
+            assert watchtower["reachable"]
```
😄

danielabrozzoni commented on tests/test_rpc.py:
```diff
@@ -1108,25 +1108,21 @@ def test_getserverstatus(revault_network, bitcoind):
-        server.stop()
+    # Alright, let's kill all the servers
+    for d in rn.daemons:
+        if d not in rn.participants():
```
Ah, cool

darosior commented:
ACK, fwiw i have been extensively testing this on signet for the past day(s).

danielabrozzoni commented on src/daemon/jsonrpc/api/mod.rs:
```diff
@@ -745,10 +746,26 @@ impl RpcApi for RpcImpl {
+        let emer_db_tx =
+            db_emer_transaction(&db_path, db_vault.id)?.ok_or(JsonRpcError::internal_error())?;
+        if !db_vault.emer_shared && emer_db_tx.psbt.unwrap_emer().is_finalizable(secp_ctx) {
+            if let Some(ref watchtowers) = revaultd.watchtowers {
```
Oh wait, so are we ok with stakeholders not having watchtowers?

darosior replied:
Yes, i see no reason to not accept it and at least for now we need it

darosior commented:
> I think it might be that pytest tries to run miradord tests as well. Did it ever occur to you?

Yeah, i specify the files to run. For instance `pytest tests/test_*.py`. But a [quick search](https://stackoverflow.com/a/41940247) would give you the more appropriate `--ignore` functionality.

darosior commented on src/daemon/jsonrpc/api/error.rs:
```diff
@@ -96,6 +103,8 @@ pub enum ErrorCode {
     /// An error internal to revault_net, generally a transport error
     TRANSPORT_ERROR = 12000,
+    /// The watchtower refused our signatures
+    WT_SIG_NACK = 13_000,
```
Because i find it much more readable and i didn't want to increase the diff with unrelated changes by modifying the others (as i did for instance to add `From` to be able to use `?` directly for some errors in #306 but didn't change all occurrences).

darosior commented on tests/test_framework/revault_network.py:
```diff
@@ -537,10 +537,11 @@ def activate_fresh_vaults(self, vaults):
             j.result(TIMEOUT)
 
-    def unvault_vaults(self, vaults, destinations, feerate):
+    def broadcast_unvaults(self, vaults, destinations, feerate):
```
Yes, but let's introduce it when we actually need it.

darosior commented on tests/test_framework/wtplugins/max_value_per_day.py:
```diff
@@ -0,0 +1,71 @@
+    - A "data_dir" entry specifying where it is going to store its 'database'.
+    - A "max_value" entry specifying the maximum value per day to enforce.
+
+It simply stores a counter which is reset to 0 after 144 blocks (assumes no reorg).
```
Yes it a random testing plugin (we won't put plugins in `test_framework/`, let alone in this repo..). Using system time includes a lot of non-determinism for the tests

darosior commented:
Added one more `wait_for_log` to try to prevent the `wt_policy` test timeout on CI which i was unable to reproduce locally.

Added a commit to run `test_watchtower` with `test_misc` to reduce CI running time since we can only have up to 4 instances running concurrently.

danielabrozzoni commented:
> Yeah, i specify the files to run. For instance pytest tests/test_*.py. But a quick search would give you the more appropriate --ignore functionality.

Ah, got it! I thought I was doing something wrong 😅 
Maybe update `tests/README.md` with the miradord download instructions and this `--ignore` functionality then :)

danielabrozzoni commented:
ACK 1c4dee062fa0d2c5ca5808ae9e2fdac2dcce97c1

darosior commented:
>  Maybe update tests/README.md with the miradord download instructions and this --ignore functionality then :)

Good call, done.

danielabrozzoni commented:
You also need:
```
diff --git a/tests/README.md b/tests/README.md
index f01f9c2..1b5c0f2 100644
--- a/tests/README.md
+++ b/tests/README.md
@@ -61,6 +61,7 @@ git submodule update --init --recursive
 cd tests/servers
 cd coordinatord && cargo build
 cd cosignerd && cargo build
+cd miradord && cargo build
 # TODO: add the other servers here when they are implemented
```

darosior commented:
:sob: done. Thanks for the sharp look!

danielabrozzoni commented:
ACK 81c532dd0f80b956f07676018f829a3ef5fff223
